### PR TITLE
Support for force-warns

### PR DESCRIPTION
**Pull request by rylev**
Implements https://github.com/rust-lang/rust/issues/85512.

This PR adds a new command line option `force-warns` which will force the provided lints to warn even if they are allowed by some other mechanism such as `#![allow(warnings)]`.

Some remaining issues:
* https://github.com/rust-lang/rust/issues/85512 mentions that `force-warns` should also be capable of taking lint groups instead of individual lints. This is not implemented. 
* If a lint has a higher warning level than `warn`, this will cause that lint to warn instead. We probably want to allow the lint to error if it is set to a higher lint and is not allowed somewhere else. 
* One test is currently ignored because it's not working - when a deny-by-default lint is allowed, it does not currently warn under `force-warns`. I'm not sure why, but I wanted to get this in before the weekend. 

r? @nikomatsakis

### CODE REVIEW

File: compiler/rustc_lint/src/levels.rs [hunk 1]
```diff
@@ -109,6 +109,11 @@ impl<'s> LintLevelsBuilder<'s> {
             }
         }
 
+        for lint_name in &sess.opts.force_warns {
+            store.check_lint_name_cmdline(sess, &lint_name, Level::Allow); // FIXME level is wrong
```
Comment (nikomatsakis) [inline compiler/rustc_lint/src/levels.rs:113]: So... I think you have a few options here to handle groups. Probably the one I like best is to do this.

```rust
match store.check_lint_name(sess, &lint_name) {
    CheckLintNameResult::Ok(lint_ids) => {
        self.sets.force_warns.extend(lint_ids);
    }
    CheckLintNameResult::NoLint(...) => {
        /* report useful error */
    }
    ...
}
```

though you could modify the `check_lint_name_cmdline` to give more appropriate errors and use `store.find_lints(&lint_name)` instead.

File: compiler/rustc_middle/src/lint.rs [hunk 2]
```diff
@@ -55,6 +60,7 @@ pub type LevelAndSource = (Level, LintLevelSource);
 pub struct LintLevelSets {
     pub list: Vec<LintSet>,
     pub lint_cap: Level,
+    pub force_warns: FxHashSet<String>,
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/lint.rs:63]: This should be changed to `FxHashSet<LintId>`

File: compiler/rustc_middle/src/lint.rs [hunk 3]
```diff
@@ -83,6 +93,11 @@ impl LintLevelSets {
         aux: Option<&FxHashMap<LintId, LevelAndSource>>,
         sess: &Session,
     ) -> LevelAndSource {
+        // Check whether we should always warn
+        if self.force_warns.contains(lint.name) {
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/lint.rs:97]: and this should be changed to `self.force_warns.contains(&LintId::of(lint))`

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Left some notes on how to make this support lint groups (I think)

File: compiler/rustc_middle/src/lint.rs [hunk 4]
```diff
@@ -346,6 +361,13 @@ pub fn struct_lint_level<'s, 'd>(
                     );
                 }
             }
+            LintLevelSource::ForceWarn(_) => {
+                sess.diag_note_once(
+                    &mut err,
+                    DiagnosticMessageId::from(lint),
+                    "Warning forced by `force-warns` commandline option",
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/lint.rs:368]:
```suggestion
                    "warning forced by `force-warns` commandline option",
```

our style guide requests lower case here

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Review by nikomatsakis — COMMENTED** [review]
Can you add an entry to the unstable book compiler flags listing:

https://github.com/rust-lang/rust/tree/master/src/doc/unstable-book/src/compiler-flags

I'm not sure if there is a SUMMARY.md to edit

**Comment by cjgillot** [comment]
Can this feature be implemented more simply as an additional lint level, similar to "forbid" but for warnings?

**Comment by rylev** [comment]
@cjgillot I _think_ it would be possible, but I'm not convinced it would be more simple. We want these lints to behave as they currently do in all situations except in the one case where we want to force them to warn. Having another lint level would force more of a cognitive burden on users and documentation maintenance.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**